File: EvacuationSystem/src/main/java/by/tyshkevich/evacuation_system/General/Services/WeatherService.java
package by.tyshkevich.evacuation_system.General.Services;

import org.json.JSONObject;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;


@Service
public class WeatherService
{
    @Value("ab3a442e3e33cb52dda464f495110325")
    private String apiKey;

    @Value("http://api.openweathermap.org/data/2.5/weather")
    private String weatherUrl;

    private final RestTemplate restTemplate = new RestTemplate();

    public JSONObject getWeatherByCoordinates(double lat, double lon)
    {
        String url = String.format("%s?lat=%f&lon=%f&units=metric&appid=%s", weatherUrl, lat, lon, apiKey);
        ResponseEntity<String> response = restTemplate.getForEntity(url, String.class);
        return new JSONObject(response.getBody());
    }

    public JSONObject getWeatherByCity(String city)
    {
        String url = String.format("%s?q=%s&units=metric&appid=%s", weatherUrl, city, apiKey);
        ResponseEntity<String> response = restTemplate.getForEntity(url, String.class);
        return new JSONObject(response.getBody());
    }

    public String determineDangerLevel(JSONObject weatherData)
    {
        JSONObject main = weatherData.getJSONObject("main");
        double temp = main.getDouble("temp");
        int pressure = main.getInt("pressure");

        double windSpeed = weatherData.getJSONObject("wind").getDouble("speed");

        String dangerLevel = "Низкий";
        if (temp > 35 || temp < -10 || windSpeed > 15 || pressure < 1000) {
            dangerLevel = "Высокий";
        } else if (temp > 30 || temp < 0 || windSpeed > 10) {
            dangerLevel = "Умеренный";
        }

        return dangerLevel;
    }
}

File: EvacSystem/EvacProject/TELEGRAM/WeatherService.cs
using System.Net.Http;
using System.Threading.Tasks;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;
using Newtonsoft.Json.Linq;
public class WeatherService
{
    private readonly HttpClient _httpClient;
    private readonly string _apiKey;
    private readonly string _weatherUrl;
    private readonly ILogger<WeatherService> _logger;

    public WeatherService(HttpClient httpClient, IConfiguration configuration, ILogger<WeatherService> logger)
    {
        _httpClient = httpClient;
        _apiKey = configuration["OpenWeatherMap:ApiKey"];
        _weatherUrl = configuration["OpenWeatherMap:WeatherUrl"];
        _logger = logger;
    }

    public async Task<JObject> GetWeatherByCoordinates(double lat, double lon)
    {
        var url = $"{_weatherUrl}?lat={lat}&lon={lon}&units=metric&appid={_apiKey}";
        _logger.LogInformation($"WeatherService: Requesting weather data for coordinates: Lat={lat}, Lon={lon}");
        try
        {
            var response = await _httpClient.GetAsync(url);
            response.EnsureSuccessStatusCode();
            var content = await response.Content.ReadAsStringAsync();
            return JObject.Parse(content);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, $"WeatherService: Failed to get weather data for coordinates: Lat={lat}, Lon={lon}");
            throw;
        }
    }

    public async Task<JObject> GetWeatherByCity(string city)
    {
        var url = $"{_weatherUrl}?q={city}&units=metric&appid={_apiKey}";
        _logger.LogInformation($"WeatherService: Requesting weather data for city: {city}");
        try
        {
            var response = await _httpClient.GetAsync(url);
            response.EnsureSuccessStatusCode();
            var content = await response.Content.ReadAsStringAsync();
            return JObject.Parse(content);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, $"WeatherService: Failed to get weather data for city: {city}");
            throw;
        }
    }

    public string DetermineDangerLevel(JObject weatherData)
    {
        try
        {
            var main = weatherData["main"];
            var temp = (double)main["temp"];
            var pressure = (int)main["pressure"];
            var windSpeed = (double)weatherData["wind"]["speed"];

            string dangerLevel = "Низкий";
            if (temp > 35 || temp < -10 || windSpeed > 15 || pressure < 1000)
            {
                dangerLevel = "Высокий";
            }
            else if (temp > 30 || temp < 0 || windSpeed > 10)
            {
                dangerLevel = "Умеренный";
            }

            _logger.LogInformation($"WeatherService: Determined danger level: {dangerLevel}, Temp={temp}, WindSpeed={windSpeed}, Pressure={pressure}");
            return dangerLevel;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "WeatherService: Error determining danger level");
            return "Неизвестно";
        }
    }
}
